homeassistant/components/binary_sensor/bayesian.py

```diff
@@ -0,0 +1,174 @@
+    device_class = get_deprecated(config, CONF_DEVICE_CLASS, CONF_SENSOR_CLASS)
+
+    async_add_devices([
+        BayesianBinarySensor(hass, name, prior, observations,
+                             probability_threshold, device_class)
+    ], True)
+    return True
+
+
+class BayesianBinarySensor(BinarySensorDevice):
+    """Representation of a Bayesian sensor."""
+
+    def __init__(self, hass, name, prior, observations, probability_threshold,
+                 device_class):
+        """Initialize the Bayesian sensor."""
+        self._hass = hass
+        self._name = name
+        self._observations = observations
+        self._probability_threshold = probability_threshold
+        self._device_class = device_class
+        self._deviation = False
+        self.prior = prior
+        self.probability = prior
+
+        self.current_obs = OrderedDict({})
+
+        self.entity_obs = {obs['entity_id']: obs for obs in self._observations}
+
+        self.watchers = {
+            'numeric_state': self.__process_numeric_state,
+            'state': self.__process_state
+        }
+
+        @callback
+        # pylint: disable=invalid-name
+        def async_threshold_sensor_state_listener(entity, old_state,
+                                                  new_state):
+            """Handle sensor state changes."""
+            if new_state.state == STATE_UNKNOWN:
+                return
+
+            entity_obs = self.entity_obs[entity]
+            platform = entity_obs['platform']
+
+            self.watchers[platform](entity_obs)
+
+            prior = self.prior
+            for obs in self.current_obs.values():
+                prior = self.__update_probability(obs, prior)
+
+            self.probability = prior
+
+            hass.async_add_job(self.async_update_ha_state, True)
+
+        for obs in self._observations:
+            entity_id = obs['entity_id']
+            async_track_state_change(hass, entity_id,
+                                     async_threshold_sensor_state_listener)
+
+    def __process_numeric_state(self, entity_observation):
+        entity = entity_observation['entity_id']
+        if condition.async_numeric_state(self._hass, entity,
+                                         entity_observation.get('below'),
+                                         entity_observation.get('above'), None,
+                                         entity_observation):
+
+            self.current_obs[entity] = entity_observation['probability']
+
+        else:
+            self.current_obs.pop(entity, None)
+
+    def __process_state(self, entity_observation):
+        entity = entity_observation['entity_id']
+        if condition.state(self._hass, entity,
+                           entity_observation.get('to_state')):
+
+            self.current_obs[entity] = entity_observation['probability']
+
+        else:
+            self.current_obs.pop(entity, None)
+
+    def __update_probability(self, prior, observation):
```

Comment:
Oops, was updating `self.probability` before a refactor.  Changed to a `@staticmethod` to keep the namespace of the class (my usual approach), if this is not preferred, I can move outside.